remotes/workqueue.go

```diff
@@ -0,0 +1,83 @@
+package remotes
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type workItem struct {
+	do   func(ctx context.Context) error
+	done chan struct{}
+	deps []chan struct{}
+}
+
+func newWorkItem(do func(ctx context.Context) error, deps ...chan struct{}) *workItem {
+	return &workItem{
+		do:   do,
+		done: make(chan struct{}),
+		deps: deps,
+	}
+}
+
+func (wi *workItem) process(ctx context.Context) error {
+	defer close(wi.done)
+	for _, d := range wi.deps {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-d:
+		}
+	}
+	select {
+	case <-ctx.Done():
+		return nil
+	default:
+		return wi.do(ctx)
+	}
+}
+
+type workQueue struct {
+	workGroup *errgroup.Group
+	todoList  chan *workItem
+	ctx       context.Context
+}
+
+//nolint: unparam
+func newWorkQueue(ctx context.Context, workerCount, todoBuffer int) *workQueue {
+	todoList := make(chan *workItem, todoBuffer)
+	workGroup, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < workerCount; i++ {
+		workGroup.Go(func() error {
+			for {
+				wi, ok := <-todoList
+				if !ok {
+					return nil
+				}
+				if err := wi.process(ctx); err != nil {
+					return err
+				}
+			}
+		})
+	}
+	return &workQueue{
+		todoList:  todoList,
+		workGroup: workGroup,
+		ctx:       ctx,
+	}
+}
+
+func (wq *workQueue) enqueue(do func(ctx context.Context) error, deps ...chan struct{}) chan struct{} {
+	wi := newWorkItem(do, deps...)
+	select {
+	case <-wq.ctx.Done():
+		return doneCh
```

Comment:
I feel like there is an `err` being lost here, but I'm not quite sure how to deal with it, it ought to bubble up somehow to the eventual caller. I think perhaps the answer is that `chan struct{}` should actually be `chan error`?

Reply:
That is something I have in mind. the underlying errgroup here reports only the first error (and swallows the following ones, triggered by the context being automatically closed). 
The problem with a chan error, is that you can read the error from it only once... which might not desirable as well. I guess reporting only the first error, and assuming the other ones should only be cancellation errors might be ok (this is the rational behind golang.org/x/sync/errgroup.Group anyway)

Reply:
Yes, I was imagining the `char error` would only ever contain a single entry, either the first `err` or `nil` on success and that the single reader would simply propagate exactly that one value, eventually it becomes (one of the candidates for) the return from the `errgroup`.

Reply:
I think we don't really care for our particular case (individual errors are always reported trough the Event channel anyway). But I guess that if we would want to keep track of the error at the workItem level, we should do something similar to context.Context (that is, having a struct with `done chan struct{}` and a `err error`).

Reply:
I'm not sure you'll need that level of complexity, you only have a single possible reader from the chan which pretty much just forwards the error up the chain. `Context` wants to handle multiple possible waiters and have the ability for them all to get the err.